### PR TITLE
Add `.asyncGenerator` and `.asyncGeneratorFunction` detection

### LINKED ISSUE #99: Add `.asyncGenerator` detection
https://jakearchibald.com/2017/async-iterators-and-generators/

### PR DESCRIPTION
**Pull request by forresst**
Fixes #99

### CODE REVIEW

File: test/test.ts [hunk 1]
```diff
@@ -572,10 +591,33 @@ test('is.generator', t => {
 	testType(t, 'generator');
 });
 
+test('is.asyncGenerator', t => {
+	testType(t, 'asyncGenerator');
+
+	const fixture = (async function * () {
```
Comment (gioragutt) [inline test/test.ts]: what's the added value of these? here you're not testing whether this is the type or not, but whether the implementation of the type "works as promised".

I think it's about the same as checking if a `string` has a `length` property...

File: test/test.ts [hunk 2]
```diff
@@ -572,10 +591,33 @@ test('is.generator', t => {
...
+});
+
 test('is.generatorFunction', t => {
 	testType(t, 'generatorFunction', ['function']);
 });
 
+test('is.asyncGeneratorFunction', t => {
+	testType(t, 'asyncGeneratorFunction', ['function']);
+
+	const fixture = async function * () {
```
Comment (gioragutt) [inline test/test.ts]: same

File: test/test.ts [hunk 3]
```diff
@@ -189,6 +190,15 @@ const types = new Map<string, Test>([
 		],
 		typename: TypeName.Generator
 	}],
+	['asyncGenerator', {
+		is: is.asyncGenerator,
+		fixtures: [
+			(async function * () {
```
Comment (gioragutt) [inline test/test.ts]: are you sure `function * ()` is the correct formatting here? with the spaces before **and** after the `*`?

if this is the formatting enforces by the tooling, ignore this comment

File: source/index.ts [hunk 4]
```diff
@@ -153,6 +157,8 @@ is.promise = (value: unknown): value is Promise<unknown> => is.nativePromise(val
 
 is.generatorFunction = isObjectOfType<GeneratorFunction>(TypeName.GeneratorFunction);
 
+is.asyncGeneratorFunction = (value: unknown): value is ((...args: any[]) => Promise<unknown>) => getObjectType(value) === TypeName.AsyncGeneratorFunction;
```
Comment (gioragutt) [inline source/index.ts]:
```typescript
((...args: any[]) => Promise<unknown>)
```
* is this the correct return type?
* such as there is a `GeneratorFunction` type (like above, L158), isn't there an `AsyncGeneratorFunction`?

File: test/test.ts (hunk 3 above)
Comment (forresst) [reply test/test.ts]: Yes, it's XO! He requests that the format be like this

File: test/test.ts (hunk 1 above)
Comment (forresst) [reply test/test.ts]: I did the same test as for asyncFunction (old L564, new L583). It's to keep the same logic.

File: test/test.ts (hunk 2 above)
Comment (forresst) [reply test/test.ts]: I did the same test as for asyncFunction (old L564, new L583). It's to keep the same logic.

File: source/index.ts (hunk 4 above)
Comment (forresst) [reply source/index.ts]: Unless I'm mistaken, it's an async function, it must return a promise like asyncFunction (old L156, New L162)

**Comment by forresst** [comment]
@gioragutt I've left comments, I stay your tuned to discuss them.

**Comment by sindresorhus** [comment]
@forresst Note https://github.com/sindresorhus/is/commit/0c3f11038638920377c44b0832ad0a89088bfe16. Could you also add this to the assertion stuff?

**Comment by forresst** [comment]
@sindresorhus It's done, I've added the assertion

**Comment by sindresorhus** [comment]
Perfect. Thank you 🙌

**Pull request merged**